rllib/algorithms/sac/sac_learner.py

```diff
@@ -0,0 +1,117 @@
+import abc
+import numpy as np
+
+from typing import Any, Dict
+
+from ray.rllib.algorithms.sac.sac import SACConfig
+from ray.rllib.core.learner.learner import Learner
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.lambda_defaultdict import LambdaDefaultDict
+from ray.rllib.utils.metrics import LAST_TARGET_UPDATE_TS, NUM_TARGET_UPDATES
+from ray.rllib.utils.schedules.scheduler import Scheduler
+from ray.rllib.utils.typing import ModuleID
+
+# Now, this is double defined: In `SACRLModule` and here. I would keep it here
+# or push it into the `Learner` as these are recurring keys in RL.
+QF_PREDS = "qf_preds"
+QF_TARGET_PREDS = "qf_target_preds"
+QF_TWIN_PREDS = "qf_twin_preds"
+
+
+# TODO (simon): Add and remove variables, if module is added or removed.
+class SACLearner(Learner):
+    @override(Learner)
+    def build(self) -> None:
+        # Store the current alpha in log form. We need it during optimization
+        # in log form.
+        self.curr_log_alpha: Dict[ModuleID, Scheduler] = LambdaDefaultDict(
+            lambda module_id: self._get_tensor_variable(
+                # Note, we want to train the temperature parameter.
+                [np.log(self.config.get_config_for_module(module_id).initial_alpha)],
+                trainable=True,
+            )
+        )
+
+        # We need to call the `super()`'s `build()` method here to have the variables
+        # for the alpha already defined.
+        super().build()
+
+        def get_target_entropy(module_id):
+            """Returns the target entropy to use for the loss.
+
+            Args:
+                module_id: Module ID for which the target entropy should be
+                    returned.
+
+            Returns:
+                Target entropy.
+            """
+            target_entropy = self.config.get_config_for_module(module_id).target_entropy
+            if target_entropy is None or target_entropy == "auto":
+                # TODO (sven): Do we always have the `config.action_space` here?
+                target_entropy = -np.prod(
+                    self._module_spec.module_specs[module_id].action_space.shape
+                )
+            return target_entropy
+
+        self.target_entropy: Dict[ModuleID, Scheduler] = LambdaDefaultDict(
```

Comment:
same: Not Schedulers, but "torch.Tensor".

Reply:
Yes, this is a `toch.Tensor`